src/Algebra/Graph/Bipartite/AdjacencyMap.hs

```diff
@@ -413,6 +426,20 @@ toBipartite m = BAM (Map.fromAscList [ (u, setRights vs) | (Left  u, vs) <- symm
         setLefts      = Set.fromAscList . lefts  . Set.toAscList
         symmetricList = Map.toAscList $ AM.adjacencyMap $ AM.symmetricClosure m
 
+-- | Construct a bipartite 'AdjacencyMap' from "Algebra.Graph.AdjacencyMap"
+-- with part identifiers obtained from a given function, adding all neeeded
+-- edges to make the graph undirected and removing all edges inside one part.
+-- Complexity: /O(m log(n))/
+--
+-- @
+-- toBipartiteWith f Algebra.Graph.AdjacencyMap.'Algebra.Graph.AdjacencyMap.empty' == 'empty'
+-- toBipartiteWith Left  x  == 'empty'
+-- toBipartiteWith Right x  == 'empty'
+-- toBipartiteWith f        == 'toBiparitite' . Algebra.Graph.AdjacencyMap.'Algebra.Graph.AdjacencyMap.gmap' f
+-- @
+toBipartiteWith :: (Ord a, Ord b, Ord c) => (a -> Either b c) -> AM.AdjacencyMap a -> AdjacencyMap b c
+toBipartiteWith f = toBipartite . AM.gmap f
```

Comment:
Wouldn't it be faster to implement `toBipartiteWith` directly and then say `toBipartite = toBipartiteWith id`?

Reply:
I'm sure it wouldn't, because values returned by an arbitrary function are not necessarily monotonic and distinct. In this case, in `toBipartite` we are going to lose some time on `Set.fromList` instead of `Set.fromAscList` and so on.

A direct implementation of `toBipartiteWith` in same fashion would be something like this:
```Haskell
toBipartiteWith :: (Ord a, Ord b, Ord c) => (a -> Either b c) -> AM.AdjacencyMap a -> AdjacencyMap b c
toBipartiteWith f m = BAM (Map.fromListWith Set.union lr)
                          (Map.fromListWith Set.union rl)
    where
        lr = do (u, vs) <- symmetricList
                case f u of
                    Left  w -> return (w, setRights vs)
                    Right _ -> []
        rl = do (u, vs) <- symmetricList
                case f u of
                    Left  _ -> []
                    Right w -> return (w, setLefts vs)
        setLefts  = Set.fromList . lefts  . map f . Set.toAscList
        setRights = Set.fromList . rights . map f . Set.toAscList
        symmetricList = Map.toAscList $ AM.adjacencyMap $ AM.symmetricClosure m
```

Again, I'm sure that `toBipartite = toBipartiteWith id` is not faster and it is very likely that it would be slower.

Reply:
I see, thanks for looking into this. I agree with your conclusions.